dataflow/run_template/main_test.py

```diff
@@ -86,6 +106,16 @@ def test_run_template_json(app):
     with app.test_request_context(json=args):
         res = main.run_template(flask.request)
         data = json.loads(res)
-        job_id = data['job']['id']
-        time.sleep(WAIT_TIME)
-        assert sp.call(['gcloud', 'dataflow', 'jobs', 'cancel', job_id]) == 0
+        dataflow_jobs_cancel(data['job']['id'])
+
+
+def dataflow_jobs_cancel(job_id):
+    # Wait time until a job can be cancelled, as a best effort.
+    # If it fails to be cancelled, the job will run for ~8 minutes.
+    time.sleep(5)  # seconds
+    request = dataflow.projects().jobs().update(
+        projectId=PROJECT,
+        jobId=job_id,
+        body={'requestedState': 'JOB_STATE_CANCELLED'}
+    )
+    request.execute()
```

Comment:
In the subprocess approach, you check for success, can you check here or is this "fire and forget"?  Probably fine either way as it's a small job.

Reply:
Yes, after closer inspection that shouldn't actually be the case. If launching the template fails it will raise an exception which will cause the test to fail.

The job can only be cancelled after it changes status from not started into ready or something like that which takes a couple seconds (usually). That's why we're sleeping for 5 seconds. However, if for some reason it takes longer and it cannot be cancelled, the job will still run for 5-8 minutes which is not a big deal, but that shouldn't cause the test to fail.

So we're cancelling as a best effort to try to minimize resources used, so I think we're fine in not checking the output of the cancel.